lsbatch/man1/bjgroup.1
.ds ]W %
.ds ]L
.nh
.TH bjgroup 1 "OpenLava Version 3.3 - Mar 2016"
.br
.SH NAME
\fBbjgroup\fR - displays information about job groups 
.SH SYNOPSIS
.BR
.PP
.PP
\fBbjgroup\fR [\fB-h\fR | \fB-V\fR] 
.SH DESCRIPTION
.BR
.PP
.PP
Displays job group information.
.PP
.SH OPTIONS
.BR
.PP
.TP
\fB-h
\fR
.IP
Prints command usage to stderr and exits. 


.TP
\fB-V
\fR
.IP
Prints OpenLava release version to stderr and exits. 
.PP
.SH DEFAULT OUTPUT
.BR
.PP
A list of job groups is displayed with the following fields:
.PP
\fBGROUP_NAME\fR
.IP
The name of the job group.
.PP
\fBNJOBS\fR
.IP
The current number of job slots in the job group.
.PP
\fBPEND\fR
.IP
The number of job slots for pending jobs in the job group.
.PP
\fBRUN\fR
.IP
The number of job slots for running jobs in the job group.
.PP
\fBSUSP\fR
.IP
The number of job slots for suspended jobs in the job group.
.PP
\fBEXITED\fR
.IP
The number of job slots for exited jobs in the job group.
.PP
\fBDONE\fR
.IP
The number of job slots for done jobs in the job group.
.PP
\fBJLIMIT\fR
.IP
The job group limit set by bgadd -L or bgmod -L. Job groups
that have no configured limits or no limit usage are indicated
by a dash (-).
.PP
.SH EXAMPLE
.PP
bjgroup
.PP
GROUP      NJOBS   PEND    RUN     SUSP    EXITED  DONE   JLIMIT
.br
/u0        1       0       1       0       0       0      3
.br
/ol        1       0       1       0       0       0      -
.br
/bbb       1       0       0       0       1       0      -
.br
/aaa       1       0       0       0       1       0      10
.br
/ol/ol     1       0       1       0       0       0      -
.br
/u0/u0     1       0       1       0       0       0      -
.PP
.SH SEE ALSO
.BR
.PP
.PP
bgadd(1), bgdel(1), bgmod(1)


NEWS
Release notes for OpenLava versions

3.3.2
=====

-- Remove the check if a group name is also a hostname as
   there can be a host on the network not part of the cluster
   with the same name.
-- Add the noheader option for bjobs.
-- Add LSB_MAX_NUM_PROCESSORS, LSB_DJOB_NUMPROC, LSB_JOB_EXECUSER
   and LSB_EFFECTIVE_RSRCREQ variables in the job environment.
-- When a multiprocessor job (-n) specifies memory rusage and span[hosts=1]
   the memory rusage applies to the entire job instead of each slots
   of the job.
-- Fix a bug in lsb.log which was causing 'bsub -K' event being
   badly interpreted by library which printed "Bad event format".
-- Implement built in resource maxslots which represents the MBD
   MXJ parameter. This allows users to sort hosts based on MXJ.
-- Implement NUMA awareness and binding of processes based on the
   machine's NUMA architecture.
-- Fix xdr error in mbatchd if there are no job groups defined
   and user run bjgroups command.
-- Implemented the job group limit 'bgadd -L n /x/y/z'. This allows
   the users to set running job limits at different level of the job group
   hierachy. If the limit is reached jobs attached to the group will be
   pending with reason "The job group limit has been reached"
-- Renamed the bjgroups command in bjgroup.
-- Implemented bgmod to change the limits of a job group.

3.3.1
======

-- Update the lsb.queues man page describing the OWNERSHIP policy.
-- Improve error message should sbatchd failed to get go ahead
   from mbatchd,
-- Fix a bug in LIM name resolution should the caller be using
   127.0.0.1
-- Add * pattern in job name.
-- Fix bug in parsing all hosts in groups in lsb.hosts.
-- Introduced new performance enhancement parameters in lsb.params.
-- Fix res/nios blocking problems preventing olmake to run
   properly.
-- Fix nios when NIOS_RWAIT_SELECT is set in lsf.conf.

3.3.0 official release
=======

3.3.pre.2
=========
-- Keep the lsb.events and lsb.acct files always open to improve
   file access performance.
-- Implement the switching of the lsb.events asynchronously in mbatchd
   child process.
-- Implemented the bjdep command. bjdep jobid prints the status of
   job dependencies for the given job.
-- Add the bjdep man page.
-- Implement NIOS_RWAIT_SELECT, if configure in lsf.conf then nios
   and ls_rwait() will not use the SIGUSR1 to detect a task that
   has finished. ls_rwait() will select() on a connection with nios
   instead and nios will send a message when a task exits.
-- Fix RES NIOS protocol bug that prevented multiple tasks running on
   the same host from being collected using ls_rwait().
-- Implemented job groups. New commands are bgadd, bgdel and bjgroups.
-- Implemented bjobs -UF unformatted bjobs output.
-- Add LSF_BINDIR in openlava startup scripts.
-- Implement bjobs -g to display only selected job groups.
-- Implement bkill -g to kill only jobs in a specific job group.

3.3.pre.1
=========
-- Fix fairshare algorithm to decrease the number of jobs per user
   only if the job has been started.
-- Fix a bug in ownership policy where wrong runtime comparison was
   performed.
-- Implement the replay of job absolute runtime for ownership policy.
-- Implement floating client feature. If the user configures in lsf.conf
   LIM_ACCEPT_FLOAT_CLIENT=y any host can submit jobs and query the
   daemons without being registered in the lsf.cluster file.

3.2.1
======
-- Fix the parsing algorithm for compact host name.
-- Fix a bug in bmgroup where by mistake the -s and -r options were dropped.

3.2.0
======
-- Fix main Makefile.am to include the m4 directory.

3.2.beta1
=========
-- Hosts member od a host group can now be specified in a compact format
   for example: (ol[1-3]).
-- Implement configure option --enable-debug. If set the compilation will
   have the -Werror option enabled. By default this option is not set.

3.2.pre
========
-- badmin ckconfig/reconfig displays configuration error and warning
   messages.
-- bsub/bmod/brestart -h displays usage information.
-- Added USER_SHARES information in lsb.users.5
-- Add lsclusters command.
-- Use qsort() to sort jobs in MBD this improves the performance
   of replay and general bsub operations.
-- Introduce a getc2() functions to read lsb.events in memory
   before parsing it.
-- Reduce the number of malloc operations when reading the
   lsb.events file.
-- Implement memory reservation if the queue has SLOT_MEMORY_RESERVE configured
   and the job requests rusage of memory. The jobs has to reserve slots
   on the hosts as well.
-- Implement a correct switching of the lim.events file.
-- Update the bmgroup command to display GROUP_SLOT and MAX_SLOTS
   sections if configured.
-- Update the bqueues command to display slot and memory reservation.
-- Allow batch hosts configured in lsb.hosts to be specified in a
   compact form, i.e. ol[1-3] the mbatchd will expand example this into 3
   hosts ol1, ol2, ol3.
-- Allow OpenLava to use cores as cpus regardless is the machines has
   hyperthreads or not. To enable this behaviour set the variable
   LIM_DEFINE_NCPUS=cores|threads in lsf.conf. Default is threads.
-- Fix bsub bug in which if user specify -n min,max the max was ignored.
-- Make fairshare configuration compatible with the LSF one.

3.1.4
======

-- Fix the problem that if the order of events in lim.events is not correct
   LIM may core dump.
-- Change the ownership of the mbatchd log file to avoid messages going
   to syslog messages file.
-- Prevent LIM from core dumping in findHost() if the input name is NULL.
-- Enable the secondary admin to run 'badmin reconfig'.
-- Fix res nios protocol to pass the correct taskid to nios.

3.1.3
======
-- Corrected the lsf.cluster man page describing LIM communication
   parameters.
-- Enable LIM communication debugging using LC_COMM class and
   LOG_INFO level.
-- Fix bug in announceMaster() which was slowing down announcement
   of slave hosts.
-- Fix a bug in the slave host timer causing slave to send load

lsbatch/cmd/bjgroup.c
/*
 * Copyright (C) 2016 David Bigagli
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lsbatch.h"
#include "cmd.h"

static void
usage(void)
{
    fprintf(stderr, "bjgroups: [ -h ] [ -V ]\n");
}

static void
print_groups(int, struct jobGroupInfo *);

int
main(int argc, char **argv)
{
    int num;
    int cc;
    struct jobGroupInfo *jgrp;

    if (lsb_init(argv[0]) < 0) {
        lsb_perror("lsb_init");
        return -1;
    }

    while ((cc = getopt(argc, argv, "hV")) != EOF) {
        switch (cc) {
            case 'V':
                fputs(_LS_VERSION_, stderr);
                return 0;
            case 'h':
                usage();
                exit(-1);
        }
    }


    jgrp = lsb_getjgrp(&num);
    if (jgrp == NULL) {
        if (lsberrno == LSBE_NO_ERROR) {
            fprintf(stderr, "\
bjgroups: No job groups yet in the system.\n");
            return -1;
        }
        fprintf(stderr, "bjgroups: %s.\n", lsb_sysmsg());
        return -1;
    }

    print_groups(num, jgrp);

    free_jobgroupinfo(num, jgrp);

    return 0;
}

/* print_groups
 */
static void
print_groups(int num, struct jobGroupInfo *jgrp)
{
    int cc;
    int i;
    int len;
    int maxlen;

    maxlen = -1;

    for (cc = 0; cc < num; cc++) {
        len = strlen(jgrp[cc].path);
        if (len > maxlen)
            maxlen = len;
    }

    printf("GROUP");

    for (cc = 0; cc < maxlen; cc++)
        printf(" ");
    /* NJOBS   PEND   RUN   SUSP   EXITED   DONE JLIMIT
     */
    printf("%-5s   %-5s   %-5s   %-5s   %-5s   %-5s  %-6s\n",
           "NJOBS", "PEND", "RUN", "SUSP", "EXITED", "DONE", "JLIMIT");

    for (cc = 0; cc < num; cc++) {

        len = maxlen - strlen(jgrp[cc].path);
        printf("%s ", jgrp[cc].path);
        /* print the path and strlen(PATH) which
         * is 4
         */
        for (i = 0; i < (len + 4); i++)
            printf(" ");

        printf("%-5d   ", jgrp[cc].counts[0]);
        printf("%-5d   ", jgrp[cc].counts[1] + jgrp[cc].counts[2]);
        printf("%-5d   ", jgrp[cc].counts[3]);
        printf("%-5d   ", jgrp[cc].counts[4] + jgrp[cc].counts[5]);
        printf("%-5d   ", jgrp[cc].counts[6]);
        printf("%-5d   ", jgrp[cc].counts[7]);
        if (jgrp[cc].max_jobs == INT32_MAX)
            printf("%-7s   ", "-");
        else
            printf("%-7d   ", jgrp[cc].max_jobs);

        printf("\n");
    }
}
